### ISSUE TITLE
Is it possible to suppress stack alignment?

### ISSUE THREAD

**Issue by db4** [Terraspace/UASM#118]
I have a lot of code that returns result from function in CF:
```asm
.CODE

option WIN64: 1

func PROC
     ; some processing
     stc
     ret
func ENDP

END
```
`UASM` assembles that as
```
func:
  0000000000000000: 48 83 EC 08        sub         rsp,8
  0000000000000004: F9                 stc
  0000000000000005: 48 83 C4 08        add         rsp,8
  0000000000000009: C3                 ret
```
and `add rsp, 8` always clears CF. How to avoid that? Looking into docs I see

> Japheth introduced W64F_STACKALIGN16 in v2.12 because he was storing locals in reverse
order: the first local was stored the last, so it wasn’t possible to have the first local aligned 16.
>
> This has been changed it so that the first local is first after the stack homing area and is
guaranteed to always be aligned to a 16 byte boundary.

Does it mean that `sub rsp,8`/ `add rsp,8` cannot be suppressed?

**Comment by john-terraspace** [comment]
Returning results via CF isn’t really ABI compliant, but if you want to you could either switch off prologue/epilogue completely for the function or look at one of the modes that would optimise it out (given the function takes no params/no locals or invokes – however that is risky as you have to always remember that state). In this case I would opt to switch off prologue/epilogue with option proc:none

**Comment by db4** [comment]
> Returning results via CF isn’t really ABI compliant

These are my private functions used internally so no ABI is involved.

> but if you want to you could either switch off prologue/epilogue completely for the function or look at one of the modes that would optimise it out

Unfortunately switching off prologue/epilogue is not an option. Many of my functions have `USES` clause that is just ignored when `option proc:none` is set:
```
.CODE

option PROC:none

func PROC PRIVATE uses rbx
     mov rbx, 0
     stc
     ret
func ENDP
```
UASM-generated code (no compilation error!):
```
func:
  0000000000000000: 48 C7 C3 00 00 00  mov         rbx,0
                    00
  0000000000000007: F9                 stc
  0000000000000008: C3                 ret
```
BTW, JWasm just did it right without any special efforts:
```
.CODE

func PROC PRIVATE uses rbx
     mov rbx, 0
     stc
     ret
func ENDP
```
JWasm-generated code:
```
func:
  0000000000000000: 53                 push        rbx
  0000000000000001: 48 C7 C3 00 00 00  mov         rbx,0
                    00
  0000000000000008: F9                 stc
  0000000000000009: 5B                 pop         rbx
  000000000000000A: C3                 ret
```
So my original question: can I somehow switch off 16-byte stack alignment and return to original JWasm behavior? If not, maybe it's worth an option?

**Comment by db4** [comment]
@john-terraspace 
> Returning results via CF isn’t really ABI compliant, but if you want to you could either switch off prologue/epilogue completely for the function or look at one of the modes that would optimise it out

I finally solved the CF problem replacing `sub rsp,n`/`add rsp,n` in prologue/epilogue with `lea rsp,[rsp-n]`/`lea rsp,[rsp+n]` that preserve flags. But `lea` CPU instruction is one byte longer than `sub`/`add`. Is it acceptable code size increase or it should be made optional? I currently switch it on with a C define but maybe an asm run-time option is a better way to go? I'm asking this because I'm going to submit a PR.

**Comment by john-terraspace** [comment]
I think some sort of switch for that would be ideal, keep the regular shorter pro/epilogue and when you want to preserve flags across the call we use another switch.

 

I want to avoid adding more options like WIN64:n as that starts all getting very messy, perhaps a new modifier on the PROC itself

 

MyProc PROC FRAME PRESERVEFLAGS USES a b c ..

 

Something like that to keep it simple on a proc by proc basis ?

**Comment by db4** [comment]
> MyProc PROC FRAME PRESERVEFLAGS USES a b c ..
Something like that to keep it simple on a proc by proc basis ?

Frankly speaking, I would like to avoid this because (at least in my case) it requires massive changes to the existing code base that is currently compiled by JWasm and that I'm going to port to UASM (both Win64 and Linux64)

what about `option FRAMEPRESERVEFLAGS:YES` or something like that? (maybe in addition to proc parameter)

**Comment by john-terraspace** [comment]
I can live with that 😊

**Comment by db4** [comment]
> > what about option FRAMEPRESERVEFLAGS:YES

> I can live with that 😊

Well, if you have a better name in your mind I would love to know it! :)

BTW, could you review https://github.com/Terraspace/UASM/pull/121? The discussed PR is based on it.

**Comment by john-terraspace** [comment]
Updated and merged #121 into 2.50

I've added this option as well now if you want to give 2.50 branch a test.
